### DISCUSSION TITLE
Python Wrapper for libdatachannel

### DISCUSSION THREAD

**Discussion by mglettig** [paullouisageneau/libdatachannel#919] [Ideas]
It would be great to get a Python Wrapper for libdatachannel. Just like the bindings for RUST and Node.js it would be great to have Python bindings.

There is a [pure Python library](https://github.com/aiortc/aiortc) available but transfer performance on resource limited systems is quite poor.

**Comment by mglettig** [comment]
Short update: I'm currently experimenting with writing a pybind11 wrapper around libdatachannel. So far I got the data channel working but I'm still working on the video track.

**Reply by paullouisageneau** [reply]
Nice job, keep me updated about the progress.

**Comment by wusspuss** [comment]
@paullouisageneau @mglettig 
As an evening project I took a different approach: make bindings to the C api, not C++. Python code is generated automatically straight from rtc.h. It tries to offer Pythonic interfaces,  allowing for usage like:
```
with PeerConnection() as pc:
    pc.gathering_state_change_callback = print
    pc.state_change_callback = lambda *args: print("State:", args)
    pc.gathering_state_callback = lambda *args: print("GatheringState:", args)
    pc.add_track(b"m=video 52313 UDP/TLS/RTP/SAVPF 96\na=mid:video")
    pc.set_local_description()
    print(
        "local description:\n",
        json.dumps({"type": "offer", "sdp": pc.local_description}),
    )
```

Some manual overrides are still needed, e.g. because it's not possible to programmatically tell if an argument is optional. Also, some names are a bit awkward, like `dc.data_channel_label` when it should ideally be just `dc.label`. These could be fixed by specifying optional args in a comment inside the rtc.h header in some consistent manner and using more consistent function naming, respectively. @paullouisageneau would such pull requests be considered? This could also help with bindings for other languages in the same manner. 

This is a quick-and-dirty evening boredom-driven development project, so `convert.tcl` can probably be much more elegant and DRY, especially if some hints/even more consistency is added to rtc.h. It is by no means complete, but it already covered more functions than I expected it to, thanks to rtc.h already being rather consistent. 

It wraps PeerConnection, DataChannel and apparently some other thing into classes, fills those classes with methods, fields for callbacks, and also wraps getters and setters in Python properties (some demonstrated in the above example). It wraps enums in Python enums, so in callbacks you get enum names, not just numbers. It translates errors in Python Exceptions.
[repo](https://github.com/wusspuss/libdatachannel-python)

**Reply by paullouisageneau** [reply]
@wusspuss Great job! I'd definitely consider PRs to ease bindings generation provided they keep acceptable backwards compatibility.
